### PR TITLE
net: lib: nrf_cloud_pgps: Fix bug in prediction set update

### PR DESCRIPTION
**Pull request by tokangas**
Fixed a bug which caused the prediction set update to fail. The failure happened when all downloaded predictions could not be written to flash sequentially.

NCSDK-25224

### CODE REVIEW

**Comment by tokangas** [comment]
I also added a couple of error checks. Previously, the library continued to the end of the download as if nothing had happened, because the error from `store_prediction()` was ignored.

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst [hunk 1]
```diff
@@ -645,6 +645,10 @@ Libraries for networking
     * The :c:func:`nrf_cloud_obj_object_add` function to reset the added object on success.
     * Custom shadow data is now passed to the application during shadow update events.
 
+* :ref:`lib_nrf_cloud_pgps` library:
+
+  * Fixed a bug in prediction set update when :kconfig:option:`CONFIG_NRF_CLOUD_PGPS_REPLACEMENT_THRESHOLD` Kconfig option was set to non-zero value.
```
Comment (peknis) [inline doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:650]:
```suggestion
  * Fixed a bug in prediction set update when the :kconfig:option:`CONFIG_NRF_CLOUD_PGPS_REPLACEMENT_THRESHOLD` Kconfig option was set to non-zero value.
```

**Review by peknis — APPROVED** [review]
Approved with a minor nit.

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_pgps.c [hunk 2]
```diff
@@ -1557,6 +1561,10 @@ static int consume_pgps_data(uint8_t pnum, const char *buf, size_t buf_len)
 					LOG_ERR("Error opening storage again:%d", err);
 					goto fail;
 				}
+			} else if (index.storage_extent < 0) {
+				LOG_ERR("Unexpected storage extent");
```
Comment (jayteemo) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_pgps.c:1565]:
```suggestion
				LOG_ERR("Unexpected storage extent:%d", index.storage_extent);
```

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_pgps_utils.c [hunk 3]
```diff
@@ -328,16 +328,21 @@ void npgps_free_block(int block)
...
 	int i;
 	int len = 0;
+	int current_block = store_block;
 
 	for (i = 0; i < num_blocks; i++) {
-		if (pool.block_used[block]) {
+		/* The first block has already been allocated for storage, so the usage check
+		 * is skipped.
+		 */
+		if (current_block != store_block &&
```
Comment (jayteemo) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_pgps_utils.c:341]:
```suggestion
		if ((current_block != store_block) &&
```

**Review by jayteemo — APPROVED** [review]

**Review by glarsennordic — APPROVED** [review]

**Comment by tokangas** [comment]
We did a test run with the fix and `npgps_get_block_extent()` still didn't return the correct number of blocks in all cases. I rewrote the function how I think it should work. It should not wrap-around, but calculate the number of blocks that can be written before the flash storage needs to be re-opened from the beginning of the flash area.

I'll be on vacation until 17th of January. @plskeggs, please remove the DNM label if the fix is fine.

**Review by plskeggs — APPROVED** [review]

**Pull request merged**